xkeycaps.man
.de EX		\"Begin example
.ne 5
.if n .sp 1
.if t .sp .5
.nf
.in +.5i
..
.de EE
.fi
.in -.5i
.if n .sp 1
.if t .sp .5
..
.TH XKeyCaps 1 "16-Aug-98" "X Version 11"
.SH NAME
xkeycaps - graphically display and edit the X keyboard mapping
.SH SYNOPSIS
.B xkeycaps
[-\fItoolkitoption\fP ...] [-\fIoption\fP ...]
.SH DESCRIPTION
The \fIxkeycaps\fP program displays a keyboard.  Moving the mouse over a 
key describes the keysyms and modifiers that that key generates.  Clicking
left on a key simulates a KeyPress event.  Clicking right on a key brings
up a menu of operations, including a command to change the keysyms that
the key generates.  This program is, in part, a graphical front-end to
.BR xmodmap (1).
.SH OPTIONS
.I xkeycaps
accepts all of the standard toolkit options, and also accepts the
following options:
.TP 4
.B \-keyboard \fIkeyboard\-name\fP\fR or \fP\fB\-kbd\fP \fIkeyboard\-name\fP
Specifies the type of keyboard to display.  There are many different computer
keyboards in the world, and \fIxkeycaps\fP must know which one you are using 
in order to function correctly.  Case does not matter when specifying a 
keyboard name.

If you're running on the console display of a Sun or HP, then \fIxkeycaps\fP
will interrogate the attached keyboard hardware directly to determine
what keyboard you're using.  But if you're running remotely, or on another
type of machine, then you must specify a keyboard somehow.
.TP 4
.B \-help
Lists the recognized values for the \fB\-keyboard\fP option.
.TP 4
.B \-gutterwidth \fInumber\fP\fR or \fP\fB\-gw\fP \fInumber\fP
Specifies the number of pixels of space to leave between each key.
.TP 4
.B \-font \fIfontname\fP
Specifies the font to use to display the keycaps.
.SH ""
The following standard X Toolkit command line arguments are commonly 
used with \fIxkeycaps\fP:
.TP 4
.B \-display \fIhost:dpy\fP
This option specifies the X server to contact.
.TP 4
.B \-geometry \fIgeometry\fP
This option specifies the preferred size and position of the window.
.TP 4
.B \-bg \fIcolor\fP
This option specifies the color to use for the background of the window.
The default is light gray.
.TP 4
.B \-fg \fIcolor\fP
This option specifies the color to use for the foreground of the window.
The default is black.
.TP 4
.B \-bw \fInumber\fP
This option specifies the width in pixels of the border surrounding 
the window.
.TP 4
.B \-xrm \fIresourcestring\fP
This option specifies a resource string to be used.  This is especially
useful for setting resources that do not have separate command line options.
.SH DISPLAY
The bottom part of the window is a drawing of a keyboard.  In the top left of
each key is printed the string which actually appears on the surface of the
key.  In the bottom right of the key is the (hexadecimal) keycode that this
key generates.

At the top of the screen are several lines of text describing the key
under the mouse (or the most recently typed key.)  These lines are:
.TP 12
.B KeyCode:
This displays the text printed on the physical key, and the keycode generated
by that key in hex, decimal, and octal.
.TP 12
.B KeySym:
This displays the set of KeySyms that this key currently generates.  
.TP 12
.B ASCII:
This displays the ASCII equivalent of this key, taking into account the
current modifier keys which are down.
.TP 12
.B Modifiers:
this displays the modifier bits which this key generates.  If a key generates
modifiers, it is a chord-key like \fIShift\fP or \fIControl\fP.
.TP 12
.B AutoRepeat:
Whether the X server claims that this key autorepeats.  I
say ``\fIclaims\fP'' because the OpenWindows X server is the only one I
have encountered for which this information is accurate.  The per-key
autorepeat flag seems to be almost-universally ignored.
.SH COMMANDS
There are several buttons in the upper left corner of the window.  They are:
.TP 4
.B Quit
Exits the program.
.TP 4
.B Select Keyboard
Pops up a dialog box from which you can change which keyboard is displayed.
The left column lists the known types of keyboards, and the right column
lists the known layouts (mappings) of those keyboards.
.TP 4
.B Type At Window
After selecting this, you are asked to click on some other window.  After
doing this, clicking on keys on the keyboard display will simulate key
events on the window you selected.  Selecting the root window or
the \fIxkeycaps\fP window turns this off.

If you are using a window manager (for example,
.BR twm (1))
in which you can lock the keyboard focus on a window and still click on 
other windows without having the focus change, then you can accomplish 
the same thing merely by focusing on another window and clicking on the 
keys in the \fIxkeycaps\fP window.
.TP 4
.B Restore Default Map
This command restores the keyboard to its default state.  If you execute
this command while displaying a keyboard which is not the type of keyboard
you are really using, your keymap will be in a nonsensical state.  There is
no way for \fIxkeycaps\fP to tell what keyboard you are using except by
taking your word for it, so don't lie.
.TP 4
.B Write Output
This command writes an \fIxmodmap\fP input file representing the current
state of the keyboard (including all of your changes) to a file in your 
home directory.  

The file will be called \fB.xmodmap-\fP\fIhostname\fP,
where \fIhostname\fP is the name of the machine you're running on.
It will warn you if the file already exists.

It prompts you with a dialog box: you can either write an \fIxmodmap\fP 
file representing the state of every key, or you can write a smaller
file which describes only the changes.

The idea is that in the appropriate startup script, you would
add a line like
.EX
xmodmap \~/.xmodmap\-\`uname \-n\`
.EE
in the appropriate init file, so that those keyboard modifications are
made each time you log in.  (If you're not sure where this command should
go, ask your system administrator, as that tends to vary from site to site.)
.PP
Clicking left on a key simulates a \fBKeyPress\fP event.  Releasing the button
simulates a \fBKeyRelease\fP event.  If you click left on a key and move the
mouse while the button is down, \fBKeyPress\fP and \fBKeyRelease\fP events 
will be simulated on every key you move the mouse over.  Think of the mouse 
as your finger: if you drag the mouse over several keys, they will go down
and up in turn.

Clicking left on a key which is associated with modifier bits (such as
Shift or Control) causes that key to ``lock'' down.  Clicking left again
releases the key.  In this way, you can generate key-chords with the mouse:
to generate \fIControl-C\fP, click left on the \fIControl\fP key, and then 
click on the \fIC\fP key.  Click on \fIControl\fP again to turn the control
modifier off.

Typing a key on the \fIreal\fP keyboard simulates 
a \fBKeyPress\fP/\fBKeyRelease\fP event pair in the same way that
clicking on a key does.

You can also combine mouse and keyboard input: for example, if you use
the mouse to select the \fIShift\fP key, and type a character, the event
that is simulated will have the \fIShift\fP modifier set.  And if you
hold down the real \fIControl\fP key, and click on the \fIC\fP key in
the window, a \fIControl-C\fP event will be generated.  (Assuming, that 
is, that your window manager does not intercept control-left-button for its 
own purposes.)

Clicking right on a key pops up a menu of commands for the given key.
They are:
.TP 4
.B Exchange Keys
After selecting this menu item, you are asked to click on another key.
That key and the key on which you brought up the menu will be exchanged.
This changes the keyboard mapping immediately.
.TP 4
.B Duplicate Key
After selecting this menu item, you are asked to click on another key.
That key will be made a copy of the key on which you brought up the menu.
That is, the two keys will generate the same set of keysyms and modifiers.
This changes the keyboard mapping immediately.
.TP 4
.B Disable Key
The key on which you brought up the menu will be made to generate no
keysyms and no modifiers.  This changes the keyboard mapping immediately.
.TP 4
.B Restore Key To Default
The key on which you brought up the menu will be restored to its default
state; no other key will be altered.  This actually changes the current
keyboard mapping.
.TP 4
.B Edit KeySyms of Key
This pops up the "Edit Key" window, which allows you to arbitrarily change
which keysyms and modifiers this key generates.

On the left side of the window is the list of the keysyms that this key
currently generates.  (A key may generate up to eight keysyms; the 
interpretation of these keysyms is described in the X protocol document,
and is summarized later in the \fBKEYSYMS AND KEYCODES\fP section of
this man page.)

The second column is a multiple-choice list of the eight modifier bits
that this key may generate.  For example, if you want a key to behave
as a ``control'' key, you should select the \fBControl\fP modifier.

The third and fourth column (the scrolling lists) are for changing the 
keysym associated with the key.  When you select a keysym-position from 
the first column, the character set and keysym will be displayed in
the scrolling lists.  Clicking on a keysym in the ``KeySym'' column will
install that keysym in the highlighted slot in the first column.

To select a keysym from a different character set, click on the character
set name in the second column.  (The \fBLatin1\fP and \fBKeyboard\fP
character sets are the most commonly used.)

At the bottom of the window are three buttons: \fBUndo\fP, \fBAbort\fP, 
and \fBOk\fP.  Clicking on \fBUndo\fP reverts the Edit Key window to the
current state of the key in question.  \fBAbort\fP closes the Edit Key
window without making any changes.  \fBOk\fP closes the Edit Key window
and installs your changes (the current keyboard mapping is modified.)
.SH KEYSYMS AND KEYCODES
To effectively edit your keyboard mapping, there are some terms you need
to be familiar with:
.TP 10
.B \fIKeyCode\fP
This is a raw scan-code that is read from the keyboard; each physical key on 
the keyboard has a different number associated with it; this mapping cannot
be changed (but that's ok.)

Generally, every keyboard has its own set of KeyCodes, which is why you will
probably need to have a different keymap for every system you use.
.TP 10
.B \fIKeySym\fP
This is a symbol which can be generated by a single press of one key on the
keyboard: for example, all letters, numbers, and punctuation are keysyms,
and so are more abstract things like ``shift'' and ``control''.

Each KeyCode (that is, key on the keyboard) is associated with certain 
KeySyms.  The KeySyms are what give the keys their semantics (and makes
the A key generate an A), not the KeyCodes.

Usually keys are associated with one or two keysyms, which correspond to
the symbols generated when the key is pressed, and when it is pressed while
the shift key is held down.  There is a special case, which is that if the
key contains only one KeySym, and it is a letter, then the Shift key does
the obvious thing that one does to letters.
.TP 10
.B \fIKeyCap\fP
Not to be confused with KeySyms, this refers to the text which is printed on
the physical keys: it is immutable (unless you repaint your keyboard...)
.TP 10
.B \fIChord\fP
This term refers to a set of two or more keys held down simultaniously (by
analogy with piano keyboards.)  All but one of the keys will generally be
Modifier Keys.  Sometimes \fIConstellation\fP is used to mean the same thing.
.TP 10
.B \fIModifier Key\fP
This is a key like shift or control, which is used to alter the interpretation
of other keys which are held down at the same time.  Generally, pressing a
modifier key without also pressing a non-modifier key does nothing.

A key is a modifier key if it has a Modifier KeySym on it.  (More specifically,
if the KeyCode of that key is associated with a Modifier KeySym.)
.TP 10
.B \fIModifier KeySym\fP
A KeySym is a modifier keysym if it has a Modifier Bit associated with it.
But, the rules are a little more complicated than that.  It's easier to 
describe by example:

For a key to behave as one expects a shift key to behave, the keycode should
have the Shift modifier bit set; and the key should generate one of the keysyms
Shift_L and Shift_R.  If either of these is not true, the key will not behave
as a shift key.

Analogously, a control key must have the Control modifier set, and use one of
the keysyms Control_L or Control_R.

This implies that if you wanted to swap your shift and control keys, it would
not be enough to simply swap their modifier bits: you must swap their keysyms
as well.  If you only swap the modifier bits, it might appear to work at first,
but other things would malfunction.

Keys like Meta (and Super, Hyper, etc.) are a bit more complicated (see below.)
.TP 10
.B \fIModifier Bit\fP
Modifier bits are attributes which certain keysyms can have.  Some modifier
bits have predefined semantics: Shift, Lock, and Control.  The remaining
modifier bits (Mod1 through Mod5) have semantics which are defined by the 
keys with which they are associated.

That is, the Control modifier means Control if it is attached to Control_L
or Control_R, and is illegal elsewhere.

But Mod1 means Meta if it is attached to Meta_L or Meta_R; but it would mean
Alt if it were attached to Alt_L or Alt_R; or Hyper with Hyper_L or Hyper_R;
and so on.  (It could not, however, be attached to Control_L, since the 
Control modifier has already spoken for those keysyms.)

If you're thinking that this is all senselessly complicated... you're right.
.SH X PROTOCOL DOCUMENT ON KEYMAPS
The following is a more precise technical explanation of how keymapping works.
This description is from the X Protocol document, and is reprinted here for
your convenience:

.RS 4
A list of KeySyms is associated with each KeyCode.  If that 
list (ignoring trailing \fBNoSymbol\fP entries) is a single KeySym ``K'',
then the list is treated as if it were the 
list \fB``K NoSymbol K NoSymbol''\fP.  
If the list (ignoring trailing \fBNoSymbol\fP entries) is a pair 
of KeySyms ``K1 K2'', then the list is treated as if it were the 
list \fB``K1 K2 K1 K2''\fP.  If the list (ignoring trailing \fBNoSymbol\fP 
entries) is a triple of KeySyms \fB``K1 K2 K3''\fP, then the list
is treated as if it were the list \fB``K1 K2 K3 NoSymbol''\fP.

The first four elements of the list are split into two groups
of KeySyms.  Group 1 contains the first and second KeySyms, 
Group 2 contains third and fourth KeySyms.  Within each group, if the
second element of the group is \fBNoSymbol\fP, then the group should be
treated as if the second element were the same as the first element, except
when the first element is an alphabetic KeySym ``K'' for which both
lowercase and uppercase forms are defined.  In that case, the group should
be treated as if the first element were the lowercase form of ``K'' and the
second element were the uppercase form of ``K''.

The standard rules for obtaining a KeySym from a KeyPress event make
use of only the Group 1 and Group 2 KeySyms; no interpretation of
other KeySyms in the list is given here.  (That is, the last four
KeySyms are unused.)

Which group to use is determined by modifier state.  Switching
between groups is controlled by the KeySym named \fBMode_switch\fP.

By attaching that KeySym to some KeyCode and attaching 
that KeyCode to any one of the modifiers \fBMod1\fP 
through \fBMod5\fP.  This modifier is called the ``group modifier''.  
For any KeyCode, Group 1 is used when the group modifier is off, 
and Group 2 is used when the group modifier is on.

Within a group, which KeySym to use is also determined by
modifier state.  The first KeySym is used when the \fBShift\fP 
and \fBLock\fP modifiers are off.  The second KeySym is used when 
the \fBShift\fP modifier is on, or when the \fBLock\fP modifier is on 
and the second KeySym is uppercase alphabetic, or when the \fBLock\fP
modifier is on and is interpreted as \fBShiftLock\fP.  Otherwise,
when the \fBLock\fP modifier is on and is interpreted as \fBCapsLock\fP,
the state of the \fBShift\fP modifier is applied first to select
a KeySym, but if that KeySym is lowercase alphabetic, then 
the corresponding uppercase KeySym is used instead.
.RE
.SH ICCCM ON THE MODIFIER MAPPING
The following is a more precise technical explanation of how modifier keys
are interpreted.  This description is from the Inter-Client Communications 
Conventions Manual, and is reprinted here for your convenience:

.RS 4
X11 supports 8 modifier bits,  of which 3 are pre-assigned to \fBShift\fP,
\fBLock\fP and \fBControl\fP.  Each modifier bit is controlled by the state
of a set of keys, and these sets are specified in a table accessed by
\fBGetModifierMapping()\fP and \fBSetModifierMapping()\fP.

A client needing to use one of the pre-assigned modifiers should assume that
the modifier table has been set up correctly to control these modifiers.
The \fILock\fP modifier should be interpreted as \fICaps Lock\fP 
or \fIShift Lock\fP according as the keycodes in its controlling set
include \fBXK_Caps_Lock\fP or \fBXK_Shift_Lock\fP.

Clients should determine the meaning of a modifier bit from the keysyms
being used to control it.

A client needing to use an extra modifier,  for example \fIMeta\fP,  should:
.RS 2

Scan the existing modifier mappings.
If it finds a modifier that contains a keycode whose set of keysyms
includes \fBXK_Meta_L\fP or \fBXK_Meta_R\fP, it should use that modifier bit.

If there is no existing modifier controlled by \fBXK_Meta_L\fP 
or \fBXK_Meta_R\fP, it should select an unused modifier bit (one with an
empty controlling set) and:
.RS 2

If there is a keycode with \fBXL_Meta_L\fP in its set of keysyms, add that
keycode to the set for the chosen modifier, then

if there is a keycode with \fBXL_Meta_R\fP in its set of keysyms, add that
keycode to the set for the chosen modifier, then

if the controlling set is still empty,  interact with the user to
select one or more keys to be \fBMeta\fP.
.RE

If there are no unused modifier bits, ask the user to take corrective action.
.RE
.RE

The above means that the \fBMod1\fP modifier does not necessarily 
mean \fIMeta\fP, although some applications (such as \fItwm\fP 
and \fIemacs 18\fP) assume that.  Any of the five unassigned modifier bits 
could mean Meta; what matters is that a modifier bit is generated by a 
keycode which is bound to the keysym \fBMeta_L\fP or \fBMeta_R\fP.

Therefore, if you want to make a ``meta'' key, the right way is to make
the keycode in question generate both a \fBMeta\fP keysym, and some
previously-unassigned modifier bit.
.SH THE MODE_SWITCH KEYSYM
In case the above didn't make sense, what the \fBMode_switch\fP keysym does 
is, basically, act as an additional kind of shift key.  If you have four 
keysyms attached to the A key, then those four keysyms will be accessed by 
the chords: A; Shift-A, Mode_Switch-A; and Mode_Switch-Shift-A, respectively.

Like any Modifier Key, for Mode_switch to function, it must have a modifier
bit attached to it.  So, select one of the bits Mod1 through Mod5 (whichever
is unused) and attach that to the Mode_switch key.
.SH THE MULTI_KEY KEYSYM
Not to be confused with Mode_switch, \fBMulti_key\fP allows the input of
multiple character sequences that represent a single character (keysym.)
A more traditional name for this keysym might have been \fICompose\fP.

The \fBMulti_key\fP keysym is \fInot\fP a modifier keysym.  That is, for
it to function properly, it should not have any modifier bits associated
with it.  This is because it is not a ``chording'' key: you do not hold
it down along with other keys.  Rather, you press Multi_key, then release
it, then press and release another key, and the two together yield a new
character.  

For example, one traditional binding would be for \fIMulti_key\fP,
followed by \fIsingle-quote\fP, followed by \fIA\fP to produce 
the \fIAacute\fP keysym.

Not all vendors support the use of the Multi_key keysym; in particular,
Digital, Sun, and HP support it, but the X Consortium does not.  (The
reason for this, I am told, is that ``Compose'' sequences are considered
obsolete; the party line is that you should be using Input Methods to do
this.)

Whether Multi_key works is a property of the \fBXt\fP library (not the
X server) so it's possible that on a single system, Multi_key might work
with some applications and not others (depending on how those applications
were compiled and linked.)

If you use Lucid Emacs or XEmacs, then you can take advantage of Multi_key 
sequences even if your version of Xt doesn't support it, by loading 
the \fIx-compose\fP library, which simulates the traditional Xt behavior.  
For more info, read the commentary at the top of the 
file "/usr/local/lib/xemacs-*/lisp/x11/x-compose.el".
.SH DEAD KEYSYMS
Dead keys work similarly Multi_key, but they are two-keystroke commands
instead of three.  For example, pressing the \fBDead_tilde\fP key, 
releasing it, then pressing the \fIA\fP key would generate the single 
keysym \fIAtilde\fP.  (They are called ``dead'' keys because they do
not, by themselves, insert characters, but instead modify the following
character typed.  But HP likes to call them ``mute'' instead of ``dead,''
no doubt to avoid frightening the children.)

Again, these are not supported by all versions of the Xt library (but
can be simulated by XEmacs.)

Also note that different vendors have different names for the dead 
keysyms.  For example: depending on your vendor, X server version, 
and the phase of the moon, you might find that the name of ``dead-tilde'' 
is Dead_Tilde, Dtilde, SunFA_Tilde, SunXK_FA_Tilde, DXK_tilde, 
DXK_tilde_accent, hpmute_asciitilde, hpXK_mute_asciitilde, or even 
XK_mute_asciitilde.  It's a mess!  You'll have to just try it and see
what works, if anything.
.SH THINGS YOU CAN'T DO
People often ask if xkeycaps or xmodmap can be used to make one key generate 
a sequence of characters.  Unfortunately, no: you can't do this sort of 
thing by manipulating the server's keymaps.  The X keyboard model just 
doesn't work that way.

The way to do such things is to set translation resources on particular 
widgets.  It has to be done on an application-by-application basis.  For 
example, here's how you would convince 
.BR xterm (1)
to insert the string \fInext\fP when you hit F17:
.EX
xterm*VT100.Translations: #override \\
    <Key>F17: string("next")
.EE
Other applications may have different mechanisms for accomplishing the
same thing, and some applications might not support it at all.  Check
the relevant man pages for specifics.

Likewise, you can't convince one key to generate another key with 
modifiers (for example, you can't make F1 behave like Ctrl-A except by 
using translations, as above.)

It is also not possible to make a keyboard key behave as a mouse button.
.SH LOSER VENDORS
Both HP and S.u.S.E. ship their systems with broken keyboard settings
by default.  They really should know better, but they don't.

As explained above, it is \fIundefined behavior\fP for one modifier bit
to be shared between two keys with dissimilar semantics.

By default, HP uses \fIMod1\fP for both \fIMeta\fP and \fIMode_switch\fP.
This means that it's impossible for a program to tell the difference
between, for example, Meta-X and Mode_switch-X.

So, to repair this mess, you need to give the Mode_switch key a different
modifier bit (\fImod2\fP, for example.)  Or, you could just remove it from
the keymap entirely.

S.u.S.E. Linux is even more screwed up than HP: whereas HP's default keymap
contains only one bug, S.u.S.E.'s default map contains three completely
different errors!

First, their default keymap has the \fIControl\fP modifier attached to both 
the \fIControl\fP key and the \fIMulti_key\fP.  This is completely crazy, 
because not only is \fIMulti_key\fP not a control key, it's not even 

xkeycaps.c
/* xkeycaps, Copyright (c) 1991, 1992, 1993, 1995, 1996, 1997, 1998 
 * Jamie Zawinski <jwz@jwz.org>
 *
 * Permission to use, copy, modify, distribute, and sell this software and its
 * documentation for any purpose is hereby granted without fee, provided that
 * the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation.  No representations are made about the suitability of this
 * software for any purpose.  It is provided "as is" without express or 
 * implied warranty.
 */

#include "version.h"

#include <stdio.h>
#include <ctype.h>
#include <X11/X.h>
#include <X11/Xlib.h>
#include <X11/Xproto.h>
#include <X11/Intrinsic.h>
#include <X11/Shell.h>
#include <X11/Xaw/Simple.h>
#include <X11/Xaw/Paned.h>
#include <X11/Xaw/Form.h>
#include <X11/Xaw/Box.h>
#include <X11/Xmu/Error.h>

#include "KbdWidgetP.h"
#include "KeyWidgetP.h"

#include "xkeycaps.h"

char *progname;
char *short_version;

#ifdef __STDC__
# define P(x)x
#else /* !__STDC__ */
# define P(x)()
#endif


static char *xkeycapsDefaults[] = {
# include "XKeyCaps_ad.h"
 0
};


static XrmOptionDescRec options [] = {
  { "-foreground",	"*Foreground",			XrmoptionSepArg, 0 },
  { "-background",	"*Background",			XrmoptionSepArg, 0 },
  { "-fg",		"*Foreground",			XrmoptionSepArg, 0 },
  { "-bg",		"*Background",			XrmoptionSepArg, 0 },
  { "-gutterwidth",	"*Keyboard.Key.gutterWidth",	XrmoptionSepArg, 0 },
  { "-gw",		"*Keyboard.Key.gutterWidth",	XrmoptionSepArg, 0 },
  { "-font",		"*Keyboard.Key.keycapFont",	XrmoptionSepArg, 0 },
  { "-fn",		"*Keyboard.Key.keycapFont",	XrmoptionSepArg, 0 },
#ifdef HAVE_XTRAP
  { "-xtrap",		"*useXTrap",			XrmoptionNoArg, "on" },
  { "-use-xtrap",	"*useXTrap",			XrmoptionNoArg, "on" },
  { "-use_xtrap",	"*useXTrap",			XrmoptionNoArg, "on" },
  { "-no-xtrap",	"*useXTrap",			XrmoptionNoArg,"off" },
  { "-no_xtrap",	"*useXTrap",			XrmoptionNoArg,"off" },
#endif
  { "-keyboard",	"*Keyboard.keyboard",		XrmoptionSepArg, 0 },
  { "-kbd",		"*Keyboard.keyboard",		XrmoptionSepArg, 0 }
};



static KeyboardWidget make_keyboard P((Widget box, Widget box2,
				       struct info_labels *info,
				       const char *name));
static void maybe_relabel_window P((KeyboardWidget keyboard));
static KeyboardWidget xkeycaps_main_loop P((KeyboardWidget keyboard,
					    XtAppContext app));


Window XSendEvent_BadWindow = 0;

static int
#ifdef __STDC__
xkeycaps_error_handler (Display *dpy, XErrorEvent *error)
#else /* ! __STDC__ */
xkeycaps_error_handler (dpy, error)
     Display *dpy;
     XErrorEvent *error;
#endif /* ! __STDC__ */
{
  switch (error->request_code)
    {
    case X_GetKeyboardMapping:
      return 0;
    case X_SendEvent:
      if (error->error_code == BadWindow && !XSendEvent_BadWindow)
	XSendEvent_BadWindow = error->resourceid;
      else
	XmuPrintDefaultErrorMessage (dpy, error, stderr);
      return 0;
    default:
      XmuPrintDefaultErrorMessage (dpy, error, stderr);
      exit (-1);
    }
}


static KeyboardWidget
#ifdef __STDC__
make_keyboard (Widget box, Widget box2, struct info_labels *info,
	       const char *name)
#else /* ! __STDC__ */
make_keyboard (box, box2, info, name)
     Widget box, box2;
     struct info_labels *info;
     const char *name;
#endif /* ! __STDC__ */
{
  Arg av[20];
  int ac = 0;
  KeyboardWidget keyboard;
  XtSetArg (av[ac], XtNfromVert, box2); ac++;
  XtSetArg (av[ac], XtNleft, XtChainLeft); ac++;
  XtSetArg (av[ac], XtNright, XtChainRight); ac++;
  XtSetArg (av[ac], XtNtop, XtChainTop); ac++;
  XtSetArg (av[ac], XtNbottom, XtChainBottom); ac++;
  if (name) XtSetArg (av[ac], "keyboard", name), ac++;
  keyboard = (KeyboardWidget)
    XtCreateManagedWidget ("keyboard", keyboardWidgetClass, box, av, ac);
  keyboard->keyboard.label_widgets = info;
  keyboard->keyboard.key_menus = make_key_menus ((KeyboardWidget *) &keyboard);
  return keyboard;
}


static void
#ifdef __STDC__
maybe_relabel_window (KeyboardWidget keyboard)
#else /* ! __STDC__ */
maybe_relabel_window (keyboard)
     KeyboardWidget keyboard;
#endif /* ! __STDC__ */
{
  /* If the user hasn't specified the -title option, set the window title
     to be something more informative.
   */
  Widget toplevel = (Widget) keyboard;
  char buf1 [100], buf2 [100];
  XrmValue value;
  char *type;
  XrmDatabase db = XtDatabase (XtDisplay (keyboard));
  char *name, *class;
  while (XtParent (toplevel)) toplevel = XtParent (toplevel);
  XtGetApplicationNameAndClass (XtDisplay (keyboard), &name, &class);
  sprintf (buf1, "%s.title", name);
  sprintf (buf2, "%s.Title", class);
  if (XrmGetResource (db, buf1, buf2, &type, &value)) return;
  sprintf (buf1, "%s; %s%s%s keyboard", class,
	   keyboard->keyboard.vendor,
	   (keyboard->keyboard.kbd_style ? " " : ""),
	   keyboard->keyboard.kbd_style);
  XStoreName (XtDisplay (toplevel), XtWindow (toplevel), buf1);
}


/* Kludge to tell the main loop to reset some variables... */
KeyboardWidget new_kbd_selected = 0;

void
#ifdef __STDC__
replace_keyboard (KeyboardWidget keyboard, const char *name)
#else /* ! __STDC__ */
replace_keyboard (keyboard, name)
     KeyboardWidget keyboard;
     const char *name;
#endif /* ! __STDC__ */
{
  /* Doing this seems to leak about 8k each time, but I really don't care. */
  Widget box, box2;
  KeyboardWidget new_kbd;
  Widget toplevel;
  Arg av[20];
  int ac = 0;
  toplevel = box = XtParent (keyboard);
  while (XtParent (toplevel)) toplevel = XtParent (toplevel);
  box2 = 0;
  XtSetArg (av[ac], XtNfromVert, &box2); ac++;
  XtGetValues ((Widget) keyboard, av, ac);
  /*  if (! box2) abort ();*  #### wtf? */
  new_kbd = make_keyboard (box, box2, keyboard->keyboard.label_widgets, name);
  XtUnmanageChild ((Widget) keyboard);
  XtDestroyWidget ((Widget) keyboard);
  new_kbd_selected = new_kbd; /* kludge... */
  XtSetKeyboardFocus (toplevel, (Widget) new_kbd);
  maybe_relabel_window (new_kbd);
}


void
#ifdef __STDC__
xkeycaps_DispatchEvent_hook (KeyboardWidget keyboard, XEvent *event)
#else /* ! __STDC__ */
xkeycaps_DispatchEvent_hook (keyboard, event)
     KeyboardWidget keyboard;
     XEvent *event;
#endif /* ! __STDC__ */
{
  /* MappingNotify and KeymapNotify events don't have an associated
     window, so there's no way to register an event-handler function
     for one of these with Xt.  Lose, lose.
   */
  if (event->xany.type == KeymapNotify)
    keyboard_handle_keymap_notify ((Widget) keyboard, 0, event, 0);
  else if (event->xany.type == MappingNotify)
    keyboard_handle_mapping_notify ((Widget) keyboard, 0, event, 0);
}


static KeyboardWidget
#ifdef __STDC__
xkeycaps_main_loop (KeyboardWidget keyboard, XtAppContext app)
#else /* ! __STDC__ */
xkeycaps_main_loop (keyboard, app)
     KeyboardWidget keyboard;
     XtAppContext app;
#endif /* ! __STDC__ */
{
  XEvent event;
  while (1)
    {
      XtAppNextEvent (app, &event);
      xkeycaps_DispatchEvent_hook (keyboard, &event);
      XtDispatchEvent (&event);

      /* Kludge to prevent us from using a freed keyboard...
	 I think we had actually been allocating the second one out
	 of the *same space* as the first one and not losing?? */
      if (new_kbd_selected)
	return new_kbd_selected;
    }
}

int
#ifdef __STDC__
main (int argc, char **argv)
#else /* ! __STDC__ */
main (argc, argv)
     int argc;
     char **argv;
#endif /* ! __STDC__ */
{
  char *class = "XKeyCaps";
  XtAppContext app;
  Widget toplevel, box, box2, buttons;
  struct info_labels *info;
  KeyboardWidget keyboard = 0;
  Arg av [20];
  int ac = 0;
  char version_buf [255];

  {
    int index;
    int i2 = strlen (version) - 3;
    char *v2 = (char *) malloc (i2);
    memcpy (v2, version+4, i2);
    version = v2;
    for (index = 0;; index++)
      if (!strncmp ("Copyright (c)", version+index, 13)) break;
    for (v2 = version+index+13; *v2; v2++)
      if (*v2 == '(') *v2 = '<';
      else if (*v2 == ')') *v2 = '>';
    memcpy (version_buf, version, i2 = index);
    version_buf [index] = '\251';
    strcpy (version_buf + index + 1, version + index + 13);
    for (;; index++)
      if (isdigit(version[index]) && isdigit(version[index+1]) &&
	  isdigit(version[index+2]) && isdigit(version[index+3]) &&
	  version[index+4] != ',')
	break;
    short_version = (char *) malloc(i2 + 14 + strlen (version + index) + 1);
    memcpy (short_version, version, i2 + 14);
    strcpy (short_version +         i2 + 14, version + index);
  }

  toplevel = XtAppInitialize (&app, class, options, XtNumber (options),
			      &argc, argv, xkeycapsDefaults, NULL, 0);

  XtGetApplicationNameAndClass (XtDisplay (toplevel), &progname, &class);

  if (argc > 1)
    {
      int ok = (!strcmp (argv[1], "-help") ||
		!strcmp (argv[1], "--help") ||
		!strcmp (argv[1], "-version") ||
		!strcmp (argv[1], "--version") ||
		!strcmp (argv[1], "-v") ||
		!strcmp (argv[1], "-h"));
      FILE *out = (ok ? stdout : stderr);
      if (!ok)
	fprintf (out, "%s: unknown option %s\n", progname, argv [1]);

      fprintf (out, "\n%s\n", short_version);
      fprintf (out, "\n\
	Use the -keyboard (or -kbd) option to specify a keyboard,\n\
	or use the `Select Keyboard' dialog box.  Use the -help\n\
	option to list recognised arguments to -kbd.  See the man\n\
	page for more details.\n\n\
	For updates, check http://www.jwz.org/xkeycaps/\n\n");

      if (!ok)
	{
	  exit(-1);
	}
      else
	{
	  fprintf(out, "Recognised keyboard names:\n\n");
	  print_kbd_choices();
	  exit(0);
	}
    }

  create_icon_pixmaps (XtDisplay (toplevel));

  ac = 0;
  box = XtCreateManagedWidget ("vertical", panedWidgetClass, toplevel, av, ac);
  ac = 0;
  XtSetArg (av[ac], XtNorientation, "horizontal"); ac++;
  box2 = XtCreateManagedWidget ("horizontal", panedWidgetClass, box, av, ac);
  buttons = make_command_widgets (box2, (Widget *) &keyboard);
  info = make_info_widget (box2, buttons);

  keyboard = make_keyboard (box, box2, info, 0);

  XtAddEventHandler ((Widget) keyboard, KeymapStateMask, False,
		     keyboard_handle_keymap_notify, 0);
  XtAddEventHandler ((Widget) keyboard, 0, True,
		     keyboard_handle_mapping_notify, 0);
  XtRealizeWidget (toplevel);

  XtSetKeyboardFocus (toplevel, (Widget) keyboard);
  maybe_relabel_window (keyboard);
  XSetErrorHandler (xkeycaps_error_handler);

  message (keyboard, "");
  message2 (keyboard, version_buf);

  if (choose_kbd_dubious_p)
    pop_up_kbd_dbox (toplevel, (void *) &keyboard, 0);

#ifdef HAVE_XTRAP
  if (keyboard->keyboard.trap_data)
    xkeycaps_xtrap_main_loop (keyboard, app);
  else
#endif
    while (1)
      keyboard = xkeycaps_main_loop (keyboard, app);
}
